### ISSUE TITLE
[ARPA] Add supplemental data to workbook uploads

### ISSUE THREAD

**Issue by cajinajr** [usdigitalresponse/usdr-gost#786]
Creating an issue for the Upload Supplemental data feature that I've been working on for engineering. Here's a [link to the 1-pager](https://docs.google.com/document/d/1A3wrYUu1mZIv-WluCW-lsWu1ETyp8U6FRW1ZIEL0flc/edit?usp=sharing) that @trumanc put together for reference.

The core issue that is being addressed is improving the workbook upload experience. One of the ways we're attempting to do this is providing a way for the uploader to include a comment or "note" (my preferred term) along with their workbook upload. Doing so would allow for uploads to outline what may have changed within the file between uploads, as well as any other thoughts they might want to include. With notes appended to the upload, others can review those to get a sense of the history of an upload series. Below are mockup screenshots (taken from my [Figma file](https://www.figma.com/file/dpMckPYtCGwPV5p85wcpTk/ARPA-Reporter?node-id=135%3A12&t=gnAVsBn3RMTIIYh4-1) for the project) that imagine what that could look like based on feedback from the partners. I believe this is ready for engineering to implement.

Backend implementation tasks:
- #895 
- #896 

Technical implementation details for front-end:
- New fields for the upload form
	- **Agency Code**
		- Presents a dropdown of Agency Codes representing Agencies available to the user, mapped to IDs of the respective agency.
		- Submits as `agencyId` in POST request form data (i.e. the user sees and selects an Agency Code from the dropdown, but the corresponding Agency ID is what is submitted to the API).
		- If the user has a default Agency, the corresponding dropdown item for that Agency should be initially preselected, but the user should still be allowed to change the selection to a different available option.
	- **Reporting Period**
		- Displays in upload the form as the name of the reporting period. This value is display-only, and should not be modifiable by the user. It should reflect the current reporting period (the same value displayed at the top of the page).
		- The ID corresponding to the displayed Reporting period should be submitted as `reportingPeriodId` in the POST request form data.
	- **Notes**
		- Freeform text area (multiline/scrollable) for the user to input plain-text notes (no rich formatting). Input should be encoded as UTF-8.
		- Submits as `notes` in the POST request form data.
- Modified schema in response from `GET /api_uploads`:
	- Upload objects (members of the top-level `uploads` array in the response body) will now contain a nullable string `notes` field, which should be displayed in the redesigned **Version History** block for each upload, as well as on the top-level Uploads page.

----

![Upload-Note1](https://user-images.githubusercontent.com/16214088/211573282-86deb414-c9a8-451a-b2df-375d50f57924.jpg)
![Upload-Note2](https://user-images.githubusercontent.com/16214088/211573296-38879330-40ef-479f-bfc7-af00e2ac4730.jpg)
![Upload-Note3](https://user-images.githubusercontent.com/16214088/211573315-6a4c2da3-5c42-4968-9814-9d057dd07dd2.jpg)

**Comment by cajinajr** [comment]
### A note about next steps ###

After last week's QA sesh, I discovered that it might be a good idea to implement a way to see each upload's note on the top level Uploads page (screenshot below)
<img width="1358" alt="Screen Shot 2023-01-10 at 9 14 34 AM" src="https://user-images.githubusercontent.com/16214088/211574900-df6dd59c-003b-4b27-a820-0b442146e255.png">

I think this would be worth exploring and would like to experiment with this. I think it would also be a good idea to continue to get feedback from users once the note feature is implemented to see if any further improvements can be made.

**Comment by mhuang-usdr** [comment]
@TylerHendrickson just wanted to make sure this one doesn't get lost as a high priority for task breakdown

**Comment by TylerHendrickson** [comment]
@mhuang-usdr and/or @trumanc  Since this feature allows the user to set the Agency explicitly in the UI, what should the behavior be in the case where an Agency ID is _also_ provided in the spreadsheet data (which is how we're currently determining this value), and the values don't match? I think it would be best to error so that the behavior is explicit, but depending on how quickly the uploaded spreadsheet data adapts to this new convention, it's possible that this wouldn't be an edge-case.

TL;DR When the form data contains an agency ID **and** the spreadsheet provides a different agency ID, what should happen?
1. Error, informing the user of the conflict
2. Prefer the Agency ID that was set in the UI
3. Prefer the Agency ID that was provided in the spreadsheet

**Comment by TylerHendrickson** [comment]
@jcomeau1 Would you be able to weigh in on the question asked in the [above comment](https://github.com/usdigitalresponse/usdr-gost/issues/786#issuecomment-1411297404)?

**Comment by jcomeau1** [comment]
Sooo...I have a couple of things on this issue.

First, the goal is to remove both the Agency and the Reporting Start and End dates from the Input template. This means no more conflict. That said, I don't know how this may affect previously uploaded spreadsheets (for closed periods). Given this, we will need to coordinate the Input Template change with the backend release.

Second, I strongly feel that the user should not have the ability to select the reporting period. There can only be one active period and they cannot change data for a prior (certified) period. The reporting period at the top of the page should be the period the upload is associated with (and while users can technically change this - which is bad - it isn't something most users even notice).

@TylerHendrickson is this helpful?

**Comment by TylerHendrickson** [comment]
FYI I switched the status of this issue back to "Eng Planning" for the meantime, since it seems there are some tweaks for me to make.

----
@jcomeau1 Definitely helpful

> First, the goal is to remove both the Agency and the Reporting Start and End dates from the Input template. This means no more conflict. That said, I don't know how this may affect previously uploaded spreadsheets (for closed periods). 

As long as previously-uploaded spreadsheets for closed periods won't get re-validated at some future point, it doesn't seem like they would be affected by a change that sets `upload.agency_id` from form POST data rather than from a spreadsheet-derived value, since previously-uploaded spreadsheets already have an `upload.agency_id` value set.

> Given this, we will need to coordinate the Input Template change with the backend release.

An alternative to coordinating rolling out the Input Template change with the backend release would be to choose one of the conflict-detection behaviors from my [previous comment](https://github.com/usdigitalresponse/usdr-gost/issues/786#issuecomment-1411297404), which would ensure that any overlap (e.g. a spreadsheet based on the old input template gets uploaded after we've rolled out this new behavior on the backend) was handled intentionally, either through our automatic reconciliation of the conflict (options 2 & 3) or by prompting the user to make a change (option 1). What do you think?

> Second, I strongly feel that the user should not have the ability to select the reporting period. There can only be one active period and they cannot change data for a prior (certified) period.

Good call – I've updated the implementation details to reflect that the "Reporting Period" field shown in the upload form should be display-only (not a dropdown) and that the `reportingPeriodId` value in the POST request form data should always reflect the reporting period at the top of the page.

**Comment by trumanc** [comment]
> When the form data contains an agency ID and the spreadsheet provides a different agency ID, what should happen?

My answer depends on the understanding that the whole point of moving the agency to the input form instead of the excel workbook is because _it is harder to update the workbook's list of agencies than it is to add it to the website's list._ Since keeping the workbook up to date is so obnoxious, I would put option 3 (use the value in the workbook) last. I don't have a strong preference between 1 and 2 (report an error vs use the agency set in the UI), though I think I lean towards reporting an error and asking the user to correct it, _but only if they can correct it by just erasing the value in the workbook_



> Second, I strongly feel that the user should not have the ability to select the reporting period

Yeah, I think this is reasonable too. I think I was imagining something that didn't necessarily allow them to select a different reporting period, but instead showed them the currently selected period so that they could easily detect if they were not in the right period before uploading. I imagined it as a dropdown selection that was grayed out and un-changeable. For some reason I feel that styling it as a greyed-out input will make it harder for a user to ignore, but some other display-only style as Ty suggests is likely just as good or better.

**Comment by TylerHendrickson** [comment]
@trumanc re

> For some reason I feel that styling it as a greyed-out input will make it harder for a user to ignore, but some other display-only style as Ty suggests is likely just as good or better.

Just to clarify: I think the implementation for this ticket should follow the mocks, which do indicate that the reporting period should be styled as a greyed-out text. Whatever mechanism the UI uses to achieve that – be it an `<input>` element or something else – is not prescribed here. The only _requirements_ that the client needs to follow with regard to reporting period are:
1. The name of the reporting period is displayed to the user in a way that's consistent with the mocks.
2. On submission, the POST form data include a `reportingPeriodId` value (which should correspond to the reporting period whose name is being displayed to the user per the previous requirement).

**Comment by TylerHendrickson** [comment]
@jcomeau1 Are you good with the updates to this ticket?

**Comment by jcomeau1** [comment]
@TylerHendrickson Looks good.

**Comment by NatHillardUSDR** [comment]
Both of the subtasks of this have now been merged, and this work should now be done! Moving to the appropriate column

**Comment by NatHillardUSDR** [comment]
For reference, here are screenshots of the relevant screens:
![Screenshot 2023-03-07 at 12 47 48 PM](https://user-images.githubusercontent.com/120750336/223506192-350540d3-3baf-4bb8-96b2-291398a43942.jpg)
![Screenshot 2023-03-07 at 12 47 51 PM](https://user-images.githubusercontent.com/120750336/223506196-23faaa7f-c98c-4c1a-84e1-17aabd6f189b.jpg)
![Screenshot 2023-03-07 at 12 47 43 PM](https://user-images.githubusercontent.com/120750336/223506200-29b17faa-f31e-4083-a84f-e3375ba2ce78.jpg)

**Comment by NatHillardUSDR** [comment]
In GOSTBusters discussion today, @jcomeau1 noted that we'll want to also make some changes to the way the spreadsheet works with these fields - I supposed that could be included as part of this umbrella? I should note that I added various validation logic to compare the value from the spreadsheet against the value from the form for agencyId and reportingPeriodId

**Comment by jcomeau1** [comment]
I believe all related functionality is now working. I'm closing this issue.
